hydra/hydra.go

```diff
@@ -0,0 +1,153 @@
+package hydra
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/axiomhq/hyperloglog"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	levelds "github.com/ipfs/go-ds-leveldb"
+	"github.com/libp2p/go-libp2p-core/network"
+	kbucket "github.com/libp2p/go-libp2p-kbucket"
+	"github.com/libp2p/hydra-booster/metrics"
+	"github.com/libp2p/hydra-booster/sybil"
+	hysybilopts "github.com/libp2p/hydra-booster/sybil/opts"
+	"github.com/multiformats/go-multiaddr"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+)
+
+type Hydra struct {
+	Sybils       []*sybil.Sybil
+	Datastore    datastore.Datastore
+	RoutingTable *kbucket.RoutingTable
+
+	hyperLock   *sync.Mutex
+	hyperlog    *hyperloglog.Sketch
+	statsTicker *time.Ticker
+}
+
+type Options struct {
+	DatastorePath string
+	GetPort       func() int
+	NSybils       int
+	BucketSize    int
+	BsCon         int
+	Relay         bool
+	Stagger       time.Duration
+}
+
+func NewHydra(options Options) (*Hydra, error) {
+	datastore, err := levelds.NewDatastore(options.DatastorePath, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create datastore: %w", err)
+	}
+
+	var nodes []*sybil.Sybil
+
+	fmt.Fprintf(os.Stderr, "Running %d DHT Instances:\n", options.NSybils)
+
+	var hyperLock sync.Mutex
+	hyperlog := hyperloglog.New()
+
+	// What is a limiter?
+	limiter := make(chan struct{}, options.BsCon)
+
+	for i := 0; i < options.NSybils; i++ {
+		time.Sleep(options.Stagger)
```

Comment:
Why do we need to stagger?

Reply:
When spawning lots of nodes I see a dial backoff error from the bootstrappers - maybe this was to mitigate this and/or not appear like a DoS attack that might trigger some nginx rules? I dunno, guessing here 😆